docs/RFCS/00000000_cascade.md

```diff
@@ -0,0 +1,1210 @@
+
+The table walk will be accomplished iteratively using a queue and perform the dependency walking algorithm described above. As the final output is a map of table IDs to table descriptors we already have a data structure that can be used for cycle detection. However, we also have to be conscious of the fact that there may be a need to iterate over the same table twice, once with an `update` and once later with a `delete` (or vice versa) due to potential dependency cycles within the graph. This adds a slight complication to the map but can be accounted for.
+
+Both memory and runtime of this function will be bounded based on the number of tables that may potentially be part of a series of cascade operations.
+
+##### `DeleteRow` and `UpdateRow`
+
+Both of `rowwriter.go`'s `DeleteRow` and `UpdateRow` will require a small split so that there are different call sites for when a cascade is desired or not. This will have no memory impact but will obviously, have a performance impact due to the cascading call itself.
+
+##### `Cascader`'s `CascadeAll`
+
+This is where the bulk of the new functionality will exist. When an operation, be it a delete or an update are executed, they will follow the normal path in their respective row writer. But they will call out to the `Cascader`. Internally, this means that the first call to `DeleteRow` or `UpdateRow` functions will perform their original write operation and then call `CascadeAll`. However, the `Cascader` may make additional calls to `DeleteRow` or `UpdateRow` on either the same or on new row writers. These additional new writes will not call `CascadeAll` again. This maintains the separation of concerns such that all foreign key relationship checking will still be done in `fk.go`; all row writes will still be performed in `rowwriter.go`; and the coordination of the cascading operations will now be done in `cascade.go`.
+
+Checking for cascading actions requires a number of steps but can be summarized thusly:
+
+1. While the queue is not empty:
+   1. Pop a table and the its changed rows off of the queue.
+   1. For each on delete and on update action:
+      1. Check the referencing table's foreign key index to see if any rows depend on the keys that were updated or deleted and generate a list of primary keys.
+      1. If any do exist, fetch the required columns needed for deleting or updating the rows that have been affected.
+      1. If one isn't cached, create a new `RowDeleter`/`RowUpdater` specifically without creating a new cascader.
+      1. Add each row that has been affected to the map of table IDs to rows to check later for referential integrity violations.
+      1. Call `DeleteRow`/`UpdateRow` for each change (without triggering another cascade call).
+      1. Enqueue the deleted/updated rows and table ID.
+1. Call each table's `RowDeleter` and `RowUpdater` foreign key constraint checker for all rows that have been updated to ensure no orphans were created.
```

Comment:
Does it make sense to have different execution paths for the DELETE and UPDATE case?